src/main/java/build/buildfarm/server/ActionCacheService.java

```diff
@@ -100,6 +103,16 @@ public void onFailure(Throwable t) {
   @Override
   public void updateActionResult(
       UpdateActionResultRequest request, StreamObserver<ActionResult> responseObserver) {
+    // A user with write access to the cache can write anything, including malicious code and
+    // binaries, which can then be returned to other users on cache lookups.  This is a security
+    // concern.  To counteract this, we allow enforcing a policy where clients cannot upload to the
+    // action cache.  In this paradigm, it is only the remote execution engine itself that populates
+    // the action cache.
+    if (policy.equals(ActionCacheAccessPolicy.READ_ONLY)) {
```

Comment:
Set a boolean in the service instance that enables writing based on this policy, rather than the equals check each time.

if (!writable) {
  <perm denied>
  return;
 }

Reply:
fixed